File: src/main/java/com/example/festimo/domain/meet/controller/ApplicationController.java
package com.example.festimo.domain.meet.controller;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;

import com.example.festimo.domain.meet.dto.ApplicationRequest;
import com.example.festimo.domain.meet.dto.ApplicationResponse;
import com.example.festimo.domain.meet.dto.LeaderApplicationResponse;
import com.example.festimo.domain.meet.service.ApplicationService;
import com.example.festimo.global.utils.jwt.JwtTokenProvider;


@RestController
@RequestMapping("/api/meet")
@Tag(name = "동행 API", description = "동행 관련 API")
public class ApplicationController {

    private final ApplicationService applicationService;
    private final JwtTokenProvider jwtTokenProvider;


    public ApplicationController(ApplicationService applicationService, JwtTokenProvider jwtTokenProvider) {
        this.applicationService = applicationService;
        this.jwtTokenProvider = jwtTokenProvider;

    }



    /**
     * 동행 신청 API
     *
     * @param request 사용자의 userId와 companyId를 포함하는 ApplicationRequest 객체
     * @return 생성된 신청 정보를 반환, HTTP 상태 코드는 CREATED(201)
     */

    @PostMapping("/applications")
    @Operation(summary = "동행 신청")
    public ResponseEntity<ApplicationResponse> apply(
            @RequestHeader("Authorization") String authorizationHeader,
            @RequestBody ApplicationRequest request) {

        String email = getEmailFromHeader(authorizationHeader);

        ApplicationResponse response = applicationService.createApplication(email, request.getCompanionId());
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    /**
     * 리더의 동행 신청 리스트 조회 API
     *
     * @param companionId 조회할 동행의 ID
    //* @param userId    조회를 시도한 유저의 ID --> 로그인 하고 바꾸기!!!
     * @return 해당 동행을 신청한 유저의 정보
     */
    @GetMapping("/companion/{companionId}")
    @Operation(summary = "리더의 동행 신청 리스트")
    public ResponseEntity<List<LeaderApplicationResponse>> getAllApplications(
            @RequestHeader("Authorization") String authorizationHeader,
            @PathVariable Long companionId
    ) {

        String email = getEmailFromHeader(authorizationHeader);

        List<LeaderApplicationResponse> responses = applicationService.getAllApplications(companionId,email);
        return ResponseEntity.status(HttpStatus.OK).body(responses);
    }

    /**
     * 리더의 동행 신청 승인 API
     *
     * @param applicationId 승인하고 싶은 신청 ID
    // * @param userId        조회를 시도한 유저의 ID --> 로그인 하고 바꾸기!!!
     */
    @PostMapping("/{applicationId}/accept")
    @Operation(summary = "리더의 동행 신청 승인")
    public ResponseEntity<Void> acceptApplication(
            @RequestHeader("Authorization") String authorizationHeader,
            @PathVariable Long applicationId
    ) {

        String email = getEmailFromHeader(authorizationHeader);

        applicationService.acceptApplication(applicationId, email);
        return ResponseEntity.ok().build();
    }

    /**
     * 리더의 동행 신청 거절 API
     *
     * @param applicationId 거절하고 싶은 신청 ID
    //  * @param userId        조회를 시도한 유저의 ID
     */
    @PatchMapping("/{applicationId}/reject")
    @Operation(summary = "리더의 동행 신청 거절")
    public ResponseEntity<Void> rejectApplication(
            @RequestHeader("Authorization") String authorizationHeader,
            @PathVariable Long applicationId) {

        String email = getEmailFromHeader(authorizationHeader);

        applicationService.rejectApplication(applicationId,email);
        return ResponseEntity.ok().build();
    }


    //Header에서 email
    private String getEmailFromHeader(String authorizationHeader) {
        String token = authorizationHeader.replace("Bearer ", "");
        return jwtTokenProvider.getEmailFromToken(token);
    }
}

File: src/main/java/com/example/festimo/domain/meet/controller/ApplicationController.kt
package com.example.festimo.domain.meet.controller

import com.example.festimo.domain.meet.dto.ApplicantReviewResponse
import com.example.festimo.domain.meet.dto.ApplicationRequest
import com.example.festimo.domain.meet.dto.ApplicationResponse
import com.example.festimo.domain.meet.dto.LeaderApplicationResponse
import com.example.festimo.domain.meet.service.ApplicationService
import com.example.festimo.global.utils.jwt.JwtTokenProvider
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable


@RestController
@RequestMapping("/api/meet")
@Tag(name = "동행 API", description = "동행 관련 API")
class ApplicationController(
    private val applicationService: ApplicationService,
    private val jwtTokenProvider: JwtTokenProvider
) {
    @PostMapping("/applications")
    @Operation(summary = "동행 신청")
    fun apply(
        @RequestHeader("Authorization") authorizationHeader: String,
        @RequestBody request: ApplicationRequest
    ): ResponseEntity<ApplicationResponse> {
        val email = getEmailFromHeader(authorizationHeader)
        val response = applicationService.createApplication(email, request.companionId)
            ?: throw IllegalStateException("Failed to create application")
        return ResponseEntity.status(HttpStatus.CREATED).body(response)
    }

    @GetMapping("/companion/{companionId}")
    @Operation(summary = "리더의 동행 신청 리스트")
    fun getAllApplications(
        @RequestHeader("Authorization") authorizationHeader: String,
        @PathVariable companionId: Long
    ): ResponseEntity<List<LeaderApplicationResponse>> {
        val email = getEmailFromHeader(authorizationHeader)
        return ResponseEntity.ok(applicationService.getAllApplications(companionId, email).orEmpty())
    }


    @PostMapping("/{applicationId}/accept")
    @Operation(summary = "리더의 동행 신청 승인")
    fun acceptApplication(
        @RequestHeader("Authorization") authorizationHeader: String,
        @PathVariable applicationId: Long
    ): ResponseEntity<Void> {
        val email = getEmailFromHeader(authorizationHeader)
        applicationService.acceptApplication(applicationId, email)
        return ResponseEntity.ok().build()
    }

    @PatchMapping("/{applicationId}/reject")
    @Operation(summary = "리더의 동행 신청 거절")
    fun rejectApplication(
        @RequestHeader("Authorization") authorizationHeader: String,
        @PathVariable applicationId: Long
    ): ResponseEntity<Void> {
        val email = getEmailFromHeader(authorizationHeader)
        applicationService.rejectApplication(applicationId, email)
        return ResponseEntity.ok().build()
    }

    @GetMapping("/{applicationId}/reviews")
    fun getApplicantReview(
        @PathVariable applicationId: Long,
        page: Int
    ): ResponseEntity<Page<ApplicantReviewResponse>> {
        val pageable: Pageable = PageRequest.of(page, 5, Sort.by("createdAt").descending())
        val reviews = applicationService.getApplicantReviews(applicationId, pageable) ?: Page.empty()
        return if (reviews.isEmpty) {
            ResponseEntity.noContent().build()
        } else {
            ResponseEntity.ok(reviews)
        }
    }

    private fun getEmailFromHeader(authorizationHeader: String): String {
        val token = authorizationHeader.replace("Bearer ", "")
        return jwtTokenProvider.getEmailFromToken(token)
    }
}
